File: interview_java/interview/dp/stock.java
package dp;

//Best Time to Buy and Sell Stock 

public class stock {
	
	// only permitted to complete at most one transaction 
	public int maxProfit_one(int[] prices) {
		if(prices == null || prices.length == 0) return 0;
		
		int max_profit_sofar = 0;
		int min_price = prices[0];
		
		for(int i = 1; i < prices.length; i++) {
			max_profit_sofar = Math.max(max_profit_sofar, prices[i] - min_price);
			min_price = Math.min(min_price, prices[i]);  // this has to be updated after updating max_profit_sofar
		}
		
		return max_profit_sofar;
	}

	// complete as many transactions as you like (i.e, buy one and sell one share of the stock multiple times). 
	// you must sell the stock before you buy again
	public int maxProfit_many(int[] prices) {
		if(prices == null || prices.length == 0) return 0;
		
		int max_profit_sofar = 0;
		
		for(int i = 1; i < prices.length; i++) {
			max_profit_sofar += (prices[i] > prices[i-1]) ? prices[i] - prices[i-1] : 0;
		}
		
		return max_profit_sofar;
	}
	
	// complete at most two transactions
	public int maxProfit_two(int[] prices) {
		if(prices == null || prices.length == 0) return 0;
		
		int n = prices.length;
		int[] max_profit_from_left = new int[n]; 
		int[] max_profit_from_right = new int[n];
		int res = 0;
		
		// from left (one transaction before i)
		max_profit_from_left[0] = 0;
		int min = prices[0];
		for(int i = 1; i < n; i++) {
			min = Math.min(min, prices[i]);
			max_profit_from_left[i] = Math.max(max_profit_from_left[i-1], prices[i]-min);
		}
		
		// from right (one transaction after i)
		max_profit_from_left[n-1] = 0;
		int max = prices[n-1];
		for(int i = n-2; i >= 0; i--) {
			max = Math.max(max, prices[i]);
			max_profit_from_right[i] = Math.max(max_profit_from_right[i+1], max-prices[i]);
		}
		
		
		for(int i = 0; i < n; i++) {
			res  = Math.max(res, max_profit_from_left[i] + max_profit_from_right[i]);
		}
		
		return res;
	}
	
	
	// complete at most k transactions
	public int maxProfit(int k, int[] prices) {
		if(prices == null || prices.length == 0) return 0;
		int n = prices.length;
		
		int[][] global_profit = new int[n][k+1];
		int[][] local_profit = new int[n][k+1];
		
		for(int i = 1; i < n; i++) {
			int diff = prices[i] - prices[i-1];
			for(int j = 1; j <= k; j++) {
				local_profit[i][j] = Math.max(global_profit[i-1][j-1] + Math.max(diff, 0), local_profit[i-1][j] + diff);
				global_profit[i][j] = Math.max(global_profit[i-1][j], local_profit[i][j]);
			}
		}
		
		return global_profit[n-1][k];
	}
	
	
	/**
	 * @param args
	 */
	public static void main(String[] args) {
		stock test = new stock();
		int[] prices = {1,-2,3,-1,6};
		int ans = test.maxProfit_one(prices);
		System.out.println(ans);
		
		ans = test.maxProfit_many(prices);
		System.out.println(ans);

		int[] prices2 = {1,3,2,9};
		ans = test.maxProfit_two(prices2);
		System.out.println(ans);

		ans = test.maxProfit(2, prices2);
		System.out.println(ans);
	}

}

File: interviewCPP/dp/stock/stock.cpp
/*
 * stock.cpp
 *
 *  Created on: Nov 21, 2016
 *      Author: songjiguo
 */

#include "stock.h"

// only permitted to complete at most one transaction
// (ie, buy one and sell one share of the stock)
int stock_buy_sell::maxProfit(vector<int> &prices)
{
	int max_profit = 0;
	int min_price = prices[0];

	for(int i = 1; (size_t)i < prices.size(); i++) {
		min_price = std::min(min_price, prices[i]);
		max_profit = std::max(max_profit, prices[i] - min_price);
	}

	return max_profit;
}

// You may complete as many transactions as you like (ie,
// buy one and sell one share of the stock multiple times).
// However, you may not engage in multiple transactions at
// the same time (ie, you must sell the stock before you buy again).
int stock_buy_sell::maxProfit_2(vector<int> &prices)
{
	int max_profit = 0;

	for(int i = 1; (size_t)i < prices.size(); i++) {
		int diff = prices[i] - prices[i-1];
		max_profit += diff > 0 ? diff:0;
	}
	return max_profit;
}


// You may complete at most two transactions
int stock_buy_sell::maxProfit_3(vector<int> &prices)
{
	if(prices.size() == 0) return 0;
	vector<vector<int>> local(prices.size(), vector<int>(3,0));
	vector<vector<int>> global(prices.size(), vector<int>(3,0));

	local[0][0] = 0;
	global[0][0] = 0;

	for(int i = 1; (size_t)i < prices.size(); i++) {
		int diff = prices[i] - prices[i-1];
		for(int j = 1; j < 3; j++) {
			local[i][j] = std::max(global[i-1][j-1] + std::max(diff,0), local[i-1][j]+diff);
			global[i][j] = std::max(local[i][j], global[i-1][j]);
		}
	}

	return global[prices.size()-1][2];

}


// You may complete at most k transactions
int stock_buy_sell::maxProfit_4(int k, vector<int> &prices)
{
	if(prices.size() == 0) return 0;
	vector<vector<int>> local(prices.size(), vector<int>(k+1,0));
	vector<vector<int>> global(prices.size(), vector<int>(k+1,0));

	local[0][0] = 0;
	global[0][0] = 0;

	for(int i = 1; (size_t)i < prices.size(); i++) {
		int diff = prices[i] - prices[i-1];
		for(int j = 1; j < k+1; j++) {
			local[i][j] = std::max(global[i-1][j-1] + std::max(diff,0), local[i-1][j]+diff);
			global[i][j] = std::max(local[i][j], global[i-1][j]);
		}
	}

	return global[prices.size()-1][k];

}



